File: frontend-comic/src/services/homeService.ts
import { ChangePasswordRequest, ComicHome, PopulerToday } from "@/types/home";
import { VipPackage, VipSubscription } from "@/types/vipPackage";
import { fetchApi, fetchApiWithFormData } from "./api";
import { UserWithNextLevel } from "@/types/user";

/**
 * Lấy danh sách truyện theo danh mục
 * @returns Danh sách truyện theo danh mục
 */
export const getComicHome = async () => {
  const response = await fetchApi<ComicHome>('/');
  return response;
}

/**
 * Lấy danh sách truyện theo danh mục
 * @param slug - Slug của danh mục
 * @param page - Trang hiện tại
 * @param limit - Số lượng truyện trên mỗi trang
 * @returns Danh sách truyện theo danh mục
 */
export const getComicBySlugCategory = async (slug: string, page: number, limit: number) => {
  const response = await fetchApi<PopulerToday[]>(`/category/${slug}?page=${page}&limit=${limit}`);
  return response;
}

/**
 * Lấy danh sách truyện theo từ khóa tìm kiếm
 * @param search - Từ khóa tìm kiếm
 * @param page - Trang hiện tại
 * @param limit - Số lượng truyện trên mỗi trang
 * @returns Danh sách truyện theo từ khóa tìm kiếm
 */
export const getComicBySearch = async (search: string, page: number, limit: number) => {
  const response = await fetchApi<PopulerToday[]>(`/search?query=${search}&page=${page}&limit=${limit}`);
  return response;
}

/**
 * Lấy danh sách truyện yêu thích của user
 * @param page - Trang hiện tại
 * @param limit - Số lượng truyện trên mỗi trang
 * @returns Danh sách truyện yêu thích của user
 */
export const getFavoritesComic = async (page: number, limit: number) => {
  const response = await fetchApi<PopulerToday[]>(`/favorites?page=${page}&limit=${limit}`);
  return response;
}

/**
 * Lấy thông tin profile của user
 * @returns Thông tin profile của user
 */
export const getProfile = async () => {
  const response = await fetchApi<UserWithNextLevel>('/profile');
  return response;
}

/**
 * Cập nhật thông tin profile
 * @param data - Dữ liệu để cập nhật thông tin profile
 * @returns true nếu thành công, false nếu không
 */
export const updateProfile = async (data: { levelTypeId: string }) => {
  const response = await fetchApi('/profile', {
    method: 'PUT',
    data: data
  });
  return response;
}

export const updateAvatar = async (file: File) => {
  const formData = new FormData();
  formData.append('avatar', file);

  const response = await fetchApiWithFormData('/change-avatar', {
    method: 'PUT',
    data: formData
  });
  return response;
}

/**
 * Thay đổi mật khẩu
 * @param data - Dữ liệu để thay đổi mật khẩu
 * @returns true nếu thành công, false nếu không
 */
export const changePassword = async (data: ChangePasswordRequest) => {
  const response = await fetchApi('/change-password', {
    method: 'POST',
    data: data
  });
  return response;
}

/**
 * Mua vip
 * @param vipPackageId - ID của vip package
 * @returns Thông tin vip package đã mua
 */
export const buyVipPackage = async (vipPackageId: string) => {
  const endpoint = `/purchase/vip`;

  const data = {
    vipPackageId: vipPackageId,
  }

  return await fetchApi<VipPackage>(endpoint, {
    method: "POST",
    data: data,
  });
}

/**
 * Lấy thông tin vip của user
 * @returns Thông tin vip của user
 */
export const getMyVipSubscription = async () => {
  const endpoint = `/my-vip`

  return await fetchApi<VipSubscription>(endpoint);
}

/**
 * Mua chương
 * @param chapterId - ID của chapter
 * @returns true nếu thành công, false nếu không
 */
export const buyChapter = async (chapterId: string) => {
  const endpoint = `/purchase/chapter`;

  const data = {
    chapterId: chapterId,
  }

  return await fetchApi<null>(endpoint, {
    method: "POST",
    data: data,
  });
}

/**
 * Gửi yêu cầu làm publisher
 * @returns true nếu thành công, false nếu không
 */
export const sendRequestPublisher = async () => {
  const endpoint = `/publisher-request`

  return await fetchApi<null>(endpoint, {
    method: "POST",
  });
}

File: backend-comic/src/main/java/com/thjvjpxx/backend_comic/service/HomeService.java
package com.thjvjpxx.backend_comic.service;

import org.springframework.web.multipart.MultipartFile;

import com.thjvjpxx.backend_comic.dto.request.ChangePassRequest;
import com.thjvjpxx.backend_comic.dto.response.BaseResponse;
import com.thjvjpxx.backend_comic.model.User;

/**
 * Service xử lý API trang chủ
 */
public interface HomeService {
    /**
     * Lấy danh sách comic
     * 
     * @return Response chứa danh sách comic
     */
    public BaseResponse<?> getHomeComic();

    /**
     * Lấy danh sách category
     * 
     * @return Response chứa danh sách category
     */
    public BaseResponse<?> getAllCategory();

    /**
     * Lấy danh sách comic theo category
     * 
     * @param slug Slug category
     * @param page Số trang
     * @param size Số lượng trong 1 trang
     * @return Response chứa danh sách comic theo category
     */
    public BaseResponse<?> getComicByCategory(String slug, int page, int size);

    /**
     * Tìm kiếm comic
     * 
     * @param query Từ khóa tìm kiếm
     * @param page  Số trang
     * @param size  Số lượng trong 1 trang
     * @return Response chứa danh sách comic theo từ khóa
     */
    public BaseResponse<?> searchComic(String query, int page, int size);

    /**
     * Lấy danh sách comic yêu thích
     * 
     * @param currentUserId ID user
     * @param page          Số trang
     * @param size          Số lượng trong 1 trang
     * @return Response chứa danh sách comic yêu thích
     */
    public BaseResponse<?> getFavorites(User user, int page, int size);

    /**
     * Lấy thông tin user
     * 
     * @param currentUserId ID user
     * @return Response chứa thông tin user
     */
    public BaseResponse<?> getProfile(User user);

    /**
     * Cập nhật thông tin user
     * 
     * @param currentUserId ID user
     * @param levelTypeId   ID level type
     * @return Response chứa thông tin user đã cập nhật
     */
    public BaseResponse<?> updateProfile(User user, String levelTypeId);

    /**
     * Thay đổi mật khẩu
     * 
     * @param currentUserId ID user
     * @param request       DTO chứa thông tin mật khẩu mới
     * @return Response chứa thông tin user đã thay đổi mật khẩu
     */
    public BaseResponse<?> changePassword(User user, ChangePassRequest request);

    /**
     * Thay đổi avatar
     * 
     * @param user User
     * @param file File ảnh
     * @return Response chứa thông tin user đã thay đổi avatar
     */
    public BaseResponse<?> changeAvatar(User user, MultipartFile file);
}
